Support batch running of SamplerFullSolveMultiApp

aeslaughter opened the pull request:
@bwspenc This is the first stage of getting the batch runs working. I still need to figure out how to 
apply mesh parameters.

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

aeslaughter committed:
Ability to run SamplerFullSolveMultiApp in batch mode

(refs #13320)

aeslaughter committed:
Create a better interface for running transfers in batch mode

aeslaughter committed:
Remove extra execute flags

aeslaughter committed:
Support for batch runs in SamplerTransfer

(refs #13320)

aeslaughter committed:
Apply clang-format

aeslaughter committed:
Restore execMultiAppTransfers parallel barrier

aeslaughter committed:
Add requirements to batch sampler tranfer tests

aeslaughter committed:
Add requirement for batch_sampler_multiapp tests

aeslaughter committed:
Support ability to use backup and restore when running batch multiapps

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/329704/) on 14a1943 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/13342/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/13342/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format dc0513c52ed7be4f53bc4eccb6e3146fb24fdb1c`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/329889/) on 31ec390 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/13342/site)

This comment will be updated on new commits.

aeslaughter committed:
Add missing header for non unity build (and fix typo)

aeslaughter committed:
Update documentation for input file location change

aeslaughter committed:
Add support for modifing MultiApp cli_args when running in batch-reset mode

aeslaughter commented:
Command-line arguments are now supported in batch mode.

aeslaughter committed:
Add batch mode support to SamplerTransientMultiApp

aeslaughter committed:
Add documentation for batch mode running of stochastic simulations

(refs #13320)

aeslaughter committed:
Update news for filename change and fix non-unity build

jiangwen84 commented on modules/stochastic_tools/src/multiapps/SamplerFullSolveMultiApp.C:
```diff
@@ -29,13 +30,118 @@ validParams<SamplerFullSolveMultiApp>()
+    _mode(getParam<MooseEnum>("mode"))
+{
+  if (_mode == "batch-reset" || _mode == "batch-restore")
+    init(n_processors());
```
Can each multiapp still be ran in parallel?

aeslaughter replied:
Are you asking if the MultiApp, when running in batch mode, can use more that one processor? If so, then no, because of this line. Is this something you need. We could perhaps make a input parameter: "processors_per_multiapp".

aeslaughter replied:
I think we will leave this for now, we can add this feature when needed.

aeslaughter committed:
Initialize _local_batch_app_index correctly

permcody commented on modules/stochastic_tools/test/tests/multiapps/batch_full_solve_multiapp/tests:
```diff
@@ -0,0 +1,56 @@
+    check_files = master_full_solve_out_runner0.e
+    check_not_exists = master_full_solve_out_runner1.e
+    max_parallel = 1
+    requirement = "The stochastic tools module shall support running sub-applications in a single batche."
```
s/batche/batch/

permcody reviewed (changes requested):
Looks pretty good - Are we set on the API for the new StochasticTransfer? It looks pretty close to the UserObject interface, which may or may not be quite what we want. There's the notion of initializing and executing, but then there's the transfer part. I guess I'm OK with finalize.

aeslaughter committed:
Remove shadow from MOOSEDocs floats

The shadows of floats diminish the text and make it diffcult to read long pages with lots of figures, so just remove the shadow and make it look like a proper document

(refs #9638)

aeslaughter committed:
Update gold for changes to plotly javascript location

aeslaughter commented on framework/src/multiapps/MultiApp.C:
```diff
@@ -696,6 +684,19 @@ MultiApp::createApp(unsigned int i, Real start_time)
 }
 
+std::string
+MultiApp::getCommandLineArguments(unsigned int local_app)
```
I can change the name of this. This function is for getting values the 'cli_args' parameter for the supplied app id. How about `getCommandLineArgumentsFromCliAargsParameter`?

aeslaughter replied:
I changed the name to be a bit more clear: `getCommandLineArgsParamHelper`

aeslaughter commented on modules/stochastic_tools/src/transfers/SamplerPostprocessorTransfer.C:
```diff
@@ -60,34 +67,70 @@ SamplerPostprocessorTransfer::initialSetup()
+void
+SamplerPostprocessorTransfer::initializeFromMultiapp()
+{
+  _local_values.resize(0);
```
I should use .clear()

aeslaughter commented on modules/stochastic_tools/src/transfers/SamplerPostprocessorTransfer.C:
```diff
@@ -60,34 +67,70 @@ SamplerPostprocessorTransfer::initialSetup()
+SamplerPostprocessorTransfer::finalizeFromMultiapp()
+{
+  // Gather the PP values from all ranks
+  _communicator.allgather(_local_values);
```
No, this can probably be gather(0,...)

aeslaughter commented on modules/stochastic_tools/src/multiapps/SamplerTransientMultiApp.C:
```diff
@@ -28,13 +29,124 @@ validParams<SamplerTransientMultiApp>()
+SamplerTransientMultiApp::solveStep(Real dt, Real target_time, bool auto_advance)
+{
+  bool last_solve_converged = true;
+  if (_mode == "batch-restore")
```
Will do.

aeslaughter committed:
Address reviewer comments

aeslaughter committed:
Fix logic problem when using gather rather than all gather

aeslaughter commented:
@bwspenc https://mooseframework.inl.gov/docs/PRs/13342/site/batch_mode.html
